### PR TITLE
News feed

### PR DESCRIPTION
**Pull request by HabaHaba05**
User which has services can post news about the vendor. 

News Feed Tab
image

DetailsPage
image

Still need to create :

- User can follow vendors and watch them post in my vendors.

- Not load all posts immediately, but create something like in Facebook, while scrolling down these post updates.

- Ability not only to write text but also to post images.

### CODE REVIEW

**Comment by HabaHaba05** [comment]
First photo from News Feed tab, second from Details.  
Also, I will add the ability not only to write text but also to post images.

File: FromLocalsToLocals/Views/Shared/Components/LovedVendorsNews/Default.cshtml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+﻿
+
+SVEIKI
```
Comment (Rangatin) [inline FromLocalsToLocals/Views/Shared/Components/LovedVendorsNews/Default.cshtml]: Why is this file needed?:)

File: FromLocalsToLocals/Models/ViewModels/PostVM.cs [hunk 2]
```diff
@@ -16,6 +17,8 @@ public class PostVM
 
         public Vendor SelectedVendor { get; set; }
 
+        public bool DisplayInDetails { get; set; } = false;
+
```
Comment (Rangatin) [inline FromLocalsToLocals/Models/ViewModels/PostVM.cs:21]: Was this `= false;` used for testing, or is it needed?

**Review by Rangatin — COMMENTED** [review]
Didn't notice any other inaccuracies, maybe because I just skimmed through the code. 
However, I wanted to ask, if this feed, like you showed in, for example, details is responsive? 
By responsive, I mean that it shrinks accordingly to shrinking page and grows when the page is made bigger.

File: FromLocalsToLocals/Models/Services/VendorService.cs [hunk 3]
```diff
@@ -64,7 +64,7 @@ public async Task DeleteAsync(Vendor vendor)
         }
         public async Task<Vendor> GetVendorAsync(int id)
         {
-            return await _context.Vendors.FirstOrDefaultAsync(m => m.ID == id);
+            return await _context.Vendors.Include(x=> x.User).FirstOrDefaultAsync(m => m.ID == id);
```
Comment (Rangatin) [inline FromLocalsToLocals/Models/Services/VendorService.cs:67]: Why `.Include(x => x.User)` is needed here?

File: FromLocalsToLocals/Views/Shared/Components/NewsFeed/Default.cshtml [hunk 4]
```diff
@@ -186,16 +186,26 @@
...
+                                                                    @x.Title
+                                                                </option>
+                                                            }
+                                                        </select>
+                                                    }
+                                                    else
+                                                    {
+                                                        <input type="hidden" asp-for="@Model.SelectedVendorTitle" value="@Model.SelectedVendor.Title" />
+                                                    }
+                                                    <input type="hidden" asp-for="@Model.PostBackUrl" value="@Html.ViewContext.HttpContext.Request.Path.Value.ToString()" />
```
Comment (Rangatin) [inline FromLocalsToLocals/Views/Shared/Components/NewsFeed/Default.cshtml:206]: Why is this input `type="hidden"`?

File: FromLocalsToLocals/Views/Vendors/Details.cshtml [hunk 5]
```diff
@@ -1,11 +1,18 @@
 ﻿@model FromLocalsToLocals.Models.Vendor
 @using Microsoft.AspNetCore.Mvc.Localization
+@inject UserManager<AppUser> userManager
 @inject IViewLocalizer _localizer
 @{
     ViewData["Title"] = _localizer["Details"];
     Layout = "~/Views/Shared/_Layout.cshtml";
+
+    var userForFeed = await  userManager.Users.Include(x=>x.Vendors).FirstOrDefaultAsync(x => x.Id == userManager.GetUserId(User));
+    if (userForFeed.Id != Model.User.Id) userForFeed = null;
```
Comment (Rangatin) [inline FromLocalsToLocals/Views/Vendors/Details.cshtml:10]: Aah, I see - you played yourself. Don't forget `{}` even if the if statement has one line;)

File: FromLocalsToLocals/Controllers/HomeController.cs [hunk 6]
```diff
@@ -121,5 +125,62 @@ public IActionResult SetLanguage(string culture, string returnUrl)
...
             return LocalRedirect(returnUrl);
         }
+    
+        [HttpGet]
+        public async Task<IActionResult> NewsFeed(PostVM model)
+        {
+
+            if (model.User == null)
+            {
+                model.User = await _userManager.Users.Include(x => x.Vendors).FirstOrDefaultAsync(x => x.Id == _userManager.GetUserId(User));
```
Comment (StauceLukas) [inline FromLocalsToLocals/Controllers/HomeController.cs:135]: Could you explain why we need use **FirstOrDefaultAsync** here? Because in other parts of our code we write **FirstOrDefault** so maybe we could implement this ?

File: FromLocalsToLocals/Views/Shared/Components/NewsFeed/Default.cshtml [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+
+            .timeline-footer a:not(.btn) {
+                color: #575d63
+            }
+
+                .timeline-footer a:not(.btn):focus,
+                .timeline-footer a:not(.btn):hover {
+                    color: #2d353c
+                }
+    </style>
```
Comment (StauceLukas) [inline FromLocalsToLocals/Views/Shared/Components/NewsFeed/Default.cshtml:169]: As @Rangatin mentioned ,  I believe we should try separate .css with html. It should be better add it to .css file and then add it throught <link>

**Review by StauceLukas — APPROVED** [review]

File: FromLocalsToLocals/Models/ViewModels/PostVM.cs (hunk 2 above)
Comment (HabaHaba05) [reply FromLocalsToLocals/Models/ViewModels/PostVM.cs:21]: Not for testing. C# sets bool default value false, so actually, it can be removed, but at least for me, it gives some clarity.

File: FromLocalsToLocals/Models/Services/VendorService.cs (hunk 3 above)
Comment (HabaHaba05) [reply FromLocalsToLocals/Models/Services/VendorService.cs:67]: It's called Eager loading. The vendor must have a foreign key to the User, and from the vendor, I will want to get this user so I use this method.

However, there is Lazy loading method also, which is better in many situations, but I done it with eager loading, if needed we can change it.

File: FromLocalsToLocals/Views/Shared/Components/LovedVendorsNews/Default.cshtml (hunk 1 above)
Comment (HabaHaba05) [reply FromLocalsToLocals/Views/Shared/Components/LovedVendorsNews/Default.cshtml]: This file is outdated (deleted).
But I needed it for testing purposes at the start :)

File: FromLocalsToLocals/Views/Shared/Components/NewsFeed/Default.cshtml (hunk 4 above)
Comment (HabaHaba05) [reply FromLocalsToLocals/Views/Shared/Components/NewsFeed/Default.cshtml:206]: It's in the form, so every time user tries to post something, a new object is created, and I needed to set the properties values of this object, this is the only way how I managed to do it.

File: FromLocalsToLocals/Controllers/HomeController.cs (hunk 6 above)
Comment (HabaHaba05) [reply FromLocalsToLocals/Controllers/HomeController.cs:135]: Hmm, there are no other differences between these two methods one is Async another no. I believe all places that can take some time of running should be async so I used it.

File: FromLocalsToLocals/Views/Shared/Components/NewsFeed/Default.cshtml (hunk 7 above)
Comment (HabaHaba05) [reply FromLocalsToLocals/Views/Shared/Components/NewsFeed/Default.cshtml:169]: Oh, literally forgot it.

**Review by Rangatin — APPROVED** [review]

File: FromLocalsToLocals/Controllers/HomeController.cs (hunk 6 above)
Comment (Rangatin) [reply FromLocalsToLocals/Controllers/HomeController.cs:135]: I think, that it is `.FirstOrDefaultAsync()` because of the await keyword. And as @HabaHaba05  mentioned, _<...> all places that can take some time of running should be async_

**Comment by HabaHaba05** [comment]
Sorry for creating this pull request so early, while it needs more work to be done.

So DON'T MERGE it, yet.

File: FromLocalsToLocals/Models/Follower.cs [hunk 8]
```diff
@@ -4,6 +4,15 @@ namespace FromLocalsToLocals.Models
 {
     public class Follower
     {
+        public Follower()
+        {}
+
```
Comment (Rangatin) [inline FromLocalsToLocals/Models/Follower.cs:9]: What purpose does the empty constructor have here? Do you call it anywhere in the code?

File: FromLocalsToLocals/ViewComponents/NewsFeed.cs [hunk 9]
```diff
@@ -25,16 +25,22 @@ public async Task<IViewComponentResult> InvokeAsync(PostVM model)
             switch (model.ActiveTab)
             {
                 case Tab.MyFeed:
-                    model.Posts = await _context.Posts.OrderBy(x => x).Include(x => x.Vendor).Join(_context.Vendors,
-                                               post => post.VendorID,
-                                               vendor => vendor.ID,
-                                               (post, vendor) => Tuple.Create(post, vendor.Image)).ToListAsync();
+                    var list = new List<Tuple<Post, byte[]>>();
+                    foreach(var follower in model.User.Folllowing)
```
Comment (Rangatin) [inline FromLocalsToLocals/ViewComponents/NewsFeed.cs:29]: I don't know if this was intentional or not, but you should fix `.Folllowing` spelling;)

File: FromLocalsToLocals/Models/Follower.cs (hunk 8 above)
Comment (HabaHaba05) [reply FromLocalsToLocals/Models/Follower.cs:9]: I believe that if remove this constructor, our program will throw an exception somewhere, where DB will try to load followers or somewhere else.

**Comment by HabaHaba05** [comment]
Everything more or less works. It can be merged, but not necessary. 
I won't make anything new till tomorrow's meeting. 

But still, some bugs are there :

- While clicking follow buttons fast, not all vendors may be followed (Not thread-safe)

- If we click follow to one of the news vendors (I'm talking about that 4 newest vendors) and then find this vendor in all list follow button doesn't change to unfollow, and if we click it we will get an alert. ( Not hard to fix, just add some js code)

In the future: 
- Not all posts load immediately.

- Somewhere user can find all the list of the vendors that he is following.
